File: android/public-app/src/main/kotlin/hu/orszembejelento/app/report/data/ReferenceRepository.kt
package hu.orszembejelento.app.report.data

import hu.orszembejelento.app.report.data.network.PublicApi
import hu.orszembejelento.app.report.domain.LineCoverage
import hu.orszembejelento.app.report.domain.RailwayLineOption
import hu.orszembejelento.app.report.domain.RailwayLinesForSettlement
import java.util.UUID

data class SettlementOption(val id: UUID, val name: String, val countyName: String?)

sealed class SettlementSearchResult {
    data class Loaded(val settlements: List<SettlementOption>) : SettlementSearchResult()
    data object Failed : SettlementSearchResult()
}

sealed class RailwayLineLookupResult {
    data class Loaded(val response: RailwayLinesForSettlement) : RailwayLineLookupResult()
    data object Failed : RailwayLineLookupResult()
}

/** Settlement search and railway-line lookup - always server-authoritative (§3, §40-41). */
class ReferenceRepository(private val api: PublicApi) {

    suspend fun searchSettlements(query: String): SettlementSearchResult {
        val response = runCatching { api.searchSettlements(query) }.getOrNull()
        val body = response?.takeIf { it.isSuccessful }?.body() ?: return SettlementSearchResult.Failed
        val options = body.mapNotNull { item ->
            runCatching { UUID.fromString(item.id) }.getOrNull()?.let { SettlementOption(it, item.name, item.countyName) }
        }
        return SettlementSearchResult.Loaded(options)
    }

    suspend fun railwayLinesOfSettlement(settlementId: UUID): RailwayLineLookupResult {
        val response = runCatching { api.railwayLinesOfSettlement(settlementId.toString()) }.getOrNull()
        val body = response?.takeIf { it.isSuccessful }?.body() ?: return RailwayLineLookupResult.Failed

        val coverage = runCatching { LineCoverage.valueOf(body.coverage) }.getOrNull() ?: return RailwayLineLookupResult.Failed
        val items = body.items.mapNotNull { item ->
            runCatching { UUID.fromString(item.id) }.getOrNull()?.let { RailwayLineOption(it, item.code, item.displayName) }
        }
        return RailwayLineLookupResult.Loaded(RailwayLinesForSettlement(coverage, items))
    }
}

File: web/public-web/src/repository/referenceRepository.ts
import { publicApi, type PublicApi } from '../api/publicApi'
import type { LineCoverage, RailwayLinesForSettlement } from '../domain/lineDecision'

export interface SettlementOption {
  readonly id: string
  readonly name: string
  readonly countyName: string | null
}

export type SettlementSearchResult =
  | { readonly kind: 'loaded'; readonly settlements: readonly SettlementOption[] }
  | { readonly kind: 'failed' }

export type RailwayLineLookupResult =
  | { readonly kind: 'loaded'; readonly response: RailwayLinesForSettlement }
  | { readonly kind: 'failed' }

/** Settlement search and railway-line lookup - always server-authoritative (§3, §40-41). */
export function createReferenceRepository(api: PublicApi = publicApi) {
  async function searchSettlements(query: string): Promise<SettlementSearchResult> {
    const result = await api.searchSettlements(query).catch(() => undefined)
    if (!result || result.status !== 200 || !result.body) return { kind: 'failed' }
    const settlements = result.body.map((item) => ({ id: item.id, name: item.name, countyName: item.countyName }))
    return { kind: 'loaded', settlements }
  }

  async function railwayLinesOfSettlement(settlementId: string): Promise<RailwayLineLookupResult> {
    const result = await api.railwayLinesOfSettlement(settlementId).catch(() => undefined)
    if (!result || result.status !== 200 || !result.body) return { kind: 'failed' }
    const coverage = result.body.coverage as LineCoverage
    if (coverage !== 'COMPLETE' && coverage !== 'PARTIAL') return { kind: 'failed' }
    const items = result.body.items.map((item) => ({ id: item.id, code: item.code, displayName: item.displayName }))
    return { kind: 'loaded', response: { coverage, items } }
  }

  return { searchSettlements, railwayLinesOfSettlement }
}

export type ReferenceRepository = ReturnType<typeof createReferenceRepository>
